### PR TITLE
[FEATURE] - DataTable Formatters

### LINKED ISSUE #13682: [FEATURE] - DataTable - Conditional Formatting
### Problem description

I think there should be better ways to do client-side and server-side conditional formatting in DataTables.
Currently, we have to use the raw HTML formatter, either specifying JS rules like [this](https://stackoverflow.com/questions/50996875/how-to-color-rows-and-or-cells-in-a-bokeh-datatable), or by specifying a column from the CDS as color like this, which works.

```python
template = f"""
    <div style="background: <%= {bg_color_column}%>;
            color: {text_color_column}">
        <%= {valueformatting} %>
    </div>
"""
formatter = HTMLTemplateFormatter(template=template)
```

### Feature description

The most straightforward approach for me is to add a new Formatter where you can specify which column to use as colors from the CDS. I did this on a branch [here](https://github.com/thomktz/bokeh/tree/add-dynamic-formatter).

Here's how it works:
```python
### Imports and Data Generation
import numpy as np
import pandas as pd
from bokeh.models import DataTable, DynamicFormatter
from bokeh.io import show

np.random.seed(42)

# Custom colors in Hex
EUROPE = "#bdfca2"
AMERICA = "#a2c4fc"
GREY = "#d8dfe6"

# Custom colors in RGB
red_color = (245, 112, 110)
green_color = (182, 232, 144)

def interpolate_color(color1, color2, t):
    """Interpolates between two RGB colors."""
    return tuple(int(a + (b - a) * t) for a, b in zip(color1, color2))

def cmap(value):
    """Gets a color from a value between 0 and 1."""
    # Determine which colors to interpolate based on value
    if value <= 0.5:
        interpolated_color = interpolate_color(red_color, (255, 255, 0), value * 2)
    else:
        interpolated_color = interpolate_color((255, 255, 0), green_color, (value - 0.5) * 2)
    # Convert RGB to Hex
    return f"#{interpolated_color[0]:02x}{interpolated_color[1]:02x}{interpolated_color[2]:02x}"


# Sample size
sample_size = 30

# Data generation
dataframe = pd.DataFrame({
    "Name": np.random.choice(["John", "Anna", "Peter", "Linda", "Thomas", "George", "Marie", "Linda"], size=sample_size),
    "Age": np.random.randint(13, 61, size=sample_size),
    "Location": np.random.choice(["New York", "Paris", "Berlin", "London", "Stockholm", "Madrid", "Washington"], size=sample_size),
    "Height": np.round(np.random.uniform(1.5, 2.0, size=sample_size), 2),
})


### DataTable with DynamicFormatters

# Solid grey color for names column
name_format = DynamicFormatter(background_color=GREY)

# Indicate if the person is over 18 years old, and set the text red if they are not
over_18 = np.where(dataframe["Age"] < 18, "red", None)
age_format = DynamicFormatter(text_color="age_color")

# Set the background color of the "Location" column based on the "Continent" column
continent = np.where(dataframe["Location"].isin(["New York", "Washington"]), AMERICA, EUROPE)
location_format = DynamicFormatter(background_color="location_color")

# Set the color of the "Height" column on a continuous scale
norm_height = (dataframe["Height"] - dataframe["Height"].min()) / (dataframe["Height"].max() - dataframe["Height"].min())
height_color = norm_height.apply(cmap)
height_format = DynamicFormatter(background_color="height_color")

# Create the DataTable
data_table = DataTable.from_data(dataframe, include_pandas_index=False, height=800)

# Add the new source columns for the formatters
data_table.source.data["age_color"] = over_18
data_table.source.data["location_color"] = continent
data_table.source.data["height_color"] = height_color

# Apply the formatters
data_table.columns[0].formatter = name_format
data_table.columns[1].formatter = age_format
data_table.columns[2].formatter = location_format
data_table.columns[3].formatter = height_format

# Show the DataTable
show(data_table)

```

**Output:**

The code shows 3 different uses of DynamicFormatter.
1 - Define a unique solid background color for the name
2 - Define a boolean rule to format the text of "Age" red, based on the value of the "Age" column itself
3 - Define a more advanced filter based on another series, to format the locations.
4 - Format the height using a continuous color-scale

image

**Sorted by height:**

image

### Potential alternatives

We could also do more of that server-side. Create 

- BooleanConditionalFormatter(value_col, color_true, color_false)
- Colormap(value_col, cmap, scaling_mode="minmax")

etc.

### Additional information

Note: My branch also introduces some changes to the recently merged DataTable.from_data method.
- `include_pandas_index`Ability to not include the DF index as a column of the CDS
- Removed the `index_position=None`, I don't think we need to enforce this. We can always use the kwargs if we want

### PR DESCRIPTION
**Pull request by thomktz**
- issues: Fixes #13682 
- tests: not yet

Still have to discuss whether this approach is worth finishing, and then make the other Formatters compatible.

### CODE REVIEW

File: bokehjs/src/lib/core/kinds.ts [hunk 1]
```diff
@@ -526,6 +527,64 @@ export namespace Kinds {
     }
   }
 
+  export class ColorSpecValue extends Kind<types.ColorSpecValue> {
```
Comment (thomktz) [inline bokehjs/src/lib/core/kinds.ts]: Not to be confused with the Glyphs ColorSpecs, which are quite different from these. I think this one is functionally more like the Python ColorSpec? i.e. either a value or a field. The names are confusing

File: bokehjs/src/lib/core/enums.ts [hunk 2]
```diff
@@ -62,6 +62,8 @@ export const FlowMode = Enum("block", "inline")
 export type FontStyle = "normal" | "italic" | "bold" | "bold italic"
 export const FontStyle = Enum("normal", "italic", "bold", "bold italic")
 
+export type FontStyleSpecValue = {value: FontStyle} | {field: string}
```
Comment (thomktz) [inline bokehjs/src/lib/core/enums.ts]: I'm not sold on any of these names btw, open to ideas

File: bokehjs/src/lib/core/kinds.ts (hunk 1 above)
Comment (bryevdv) [reply bokehjs/src/lib/core/kinds.ts]: AFAIK the existing `ColorSpec` should also be what it wanted here? i.e. it is a data spec that is specialized to accept scalar color values, or refer to color values from a CDS column. What issues did you run into trying to use the existing `ColorSpec`? 

cc @mattpap for any thoughts or guidance, or if there are reasons existing DataSpecs couldn't be re-used in other contexts like this.

File: bokehjs/src/lib/core/enums.ts (hunk 2 above)
Comment (bryevdv) [reply bokehjs/src/lib/core/enums.ts]: Usually we'd go with `FontStyleSpec` and similar to `ColorSpec` this should already exist:

https://github.com/bokeh/bokeh/blob/branch-3.4/bokehjs/src/lib/core/properties.ts#L739

So I think the next task is to figure out how to leverage the existing dataspecs (or understand why they are not sufficient)

File: bokehjs/src/lib/core/kinds.ts (hunk 1 above)
Comment (thomktz) [reply bokehjs/src/lib/core/kinds.ts]: @bryevdv I couldn't manage to get them working at first, still unfamiliar with JS and the bokeh codebase, but I managed to use the default specs in latest commit

File: bokehjs/src/lib/core/enums.ts (hunk 2 above)
Comment (thomktz) [reply bokehjs/src/lib/core/enums.ts]: Done in last commit

File: bokehjs/src/lib/models/widgets/tables/cell_formatters.ts [hunk 3]
```diff
@@ -56,39 +57,69 @@ export class StringFormatter extends CellFormatter {
...
         text.style.fontStyle = "italic"
+        text.style.fontWeight = "bold"
         break
     }
 
-    text.style.textAlign = text_align
+    // Text align
+    if (isValue(text_align)) {
+      text.style.textAlign = text_align.value
+    } else if (isField(text_align)) {
```
Comment (thomktz) [inline bokehjs/src/lib/models/widgets/tables/cell_formatters.ts:107]: I think I could just use `else` here, but I'm still checking `isField` for safety. Insights? Can it ever be given as anything else than a Value or Field?

File: bokehjs/test/codebase/size.ts [hunk 4]
```diff
@@ -8,7 +8,7 @@ const LIMITS = new Map([
   // es2020
   ["js/bokeh.min.js",                1000],
   ["js/bokeh-widgets.min.js",         350],
-  ["js/bokeh-tables.min.js",          350],
+  ["js/bokeh-tables.min.js",          450],
```
Comment (thomktz) [inline bokehjs/test/codebase/size.ts]: Had to increase this to pass test. I think the inclusion of the ColorMaps means it has to have the color palettes in there?

File: bokehjs/src/lib/models/widgets/tables/cell_formatters.ts [hunk 5]
```diff
@@ -2,14 +2,15 @@ import tz from "timezone"
...
-import type {Color} from "core/types"
-import {FontStyle, TextAlign, RoundingFunction} from "core/enums"
+import {isField, isValue} from "core/vectorization"
+import {RoundingFunction} from "core/enums"
 import {isNumber, isString} from "core/util/types"
 import {to_fixed} from "core/util/string"
-import {color2css} from "core/util/color"
+import {color2css, rgba2css} from "core/util/color"
 import {Model} from "../../../model"
+import {ColorMapper} from "api"
```
Comment (bryevdv) [inline bokehjs/src/lib/models/widgets/tables/cell_formatters.ts]: Generally speaking, nothing inside the library proper should use or import from `api`, it is only for external users.   So this should be updated to use the in-library location.

File: bokehjs/test/codebase/size.ts (hunk 4 above)
Comment (bryevdv) [reply bokehjs/test/codebase/size.ts]: It's possible this is due to importing from `api`, but @mattpap can also comment for certain about this.

File: bokehjs/src/lib/models/widgets/tables/cell_formatters.ts (hunk 3 above)
Comment (bryevdv) [reply bokehjs/src/lib/models/widgets/tables/cell_formatters.ts:107]: I don't think so but cc @mattpap to comment. If we leave the check in we'd also generally add `else { unreachable() }` at the end for good measure.

**Comment by bryevdv** [comment]
@thomktz I've left a few comments, I've also added @mattpap for review since he is most heavily involved in BokehJS development these days. He may have some suggestions and guidance for adding other kinds of tests, or details of the implementation.

Some other things that might also be nice, depending on your availability to devote more time:

* apply this new feature to some of the existing examples that use data tables, especially any already in the gallery
* find somewhere in the user guide to make mention of this capability cc @tcmetzger for thoughts

We can make issues for follow-on if you don't have bandwidth for these. 

Lastly @thomktz would you like to be added to @bokeh/dev ? This is really a terrific first PR for diving in to BokehJS, and certainly reason to add you in case you'd are able to continue being involved. 

This is going to be a great feature to show off in a release blog!

File: bokehjs/test/codebase/size.ts (hunk 4 above)
Comment (thomktz) [reply bokehjs/test/codebase/size.ts]: Changed the import and the size went back down

**Comment by thomktz** [comment]
I do have some time to make thorough examples, docs and tests for this, with guidance.
Thanks @bryevdv, I'd love to join :) Mostly to continue working on tables, for now

Regarding examples, I have modified these two:

integration/widgets/data_table_customization.py old/new
image

interaction/tools/point_draw.py old/new
image

File: bokehjs/src/lib/models/widgets/tables/cell_formatters.ts [hunk 6]
```diff
@@ -56,39 +58,81 @@ export class StringFormatter extends CellFormatter {
...
     const text = div(value == null ? "" : `${value}`)
-    switch (font_style) {
+
+    // Font style
+    let resolvedFontStyle
+    if (isValue(font_style)) {
+      resolvedFontStyle = font_style.value
+    } else if (isField(font_style)) {
+      resolvedFontStyle = _dataContext[font_style.field]
+    }
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/tables/cell_formatters.ts:86]: This is unfortunate, because this should be deferred to vectorized properties to figure out how to resolve this. Especially to handle ndarrays properly, among other issues. Having access to the data source, we could write something like this: 
```ts
this.properties.font_style.uniform(data_source).get(index)
```
though this would need to be precomputed, perhaps as some kind of virtual column in table's data provider.

File: bokehjs/src/lib/models/widgets/tables/cell_formatters.ts [hunk 7]
```diff
@@ -56,39 +58,81 @@ export class StringFormatter extends CellFormatter {
...
+
+    switch (resolvedFontStyle) {
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/tables/cell_formatters.ts]: Our APIs use snake case, so:
```suggestion
    switch (resolved_font_style) {
```
and in all other cases.

File: bokehjs/src/lib/models/widgets/tables/cell_formatters.ts (hunk 3 above)
Comment (mattpap) [reply bokehjs/src/lib/models/widgets/tables/cell_formatters.ts:107]: The is a third option `isExpr()` which isn't handled here. I would recommend using at least something like:
```ts
} else if (isExpr(text_align)) {
  // TODO not implemented
} else {
  unrechable()
}
```

File: examples/integration/widgets/data_table_formatting.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
+from datetime import datetime
```
Comment (mattpap) [inline examples/integration/widgets/data_table_formatting.py]: New examples like this should go under `examples/interaction/widgets/`. `examples/integration` are legacy integration tests, which are currently handled in bokehjs' test suite.

File: examples/integration/widgets/data_table_customization.py [hunk 9]
```diff
@@ -3,22 +3,31 @@
                           NumberFormatter, StringFormatter, TableColumn)
 from bokeh.sampledata.periodic_table import elements
 
+
+def get_text_color(hex_color):
+    """Get a text color with high contrast v.s. the background color."""
+    r, g, b = int(hex_color[1:3], 16), int(hex_color[3:5], 16), int(hex_color[5:7], 16)
+    luminance = (0.299 * r + 0.587 * g + 0.114 * b) / 255
+    return '#000000' if luminance > 0.5 else '#ffffff'
```
Comment (mattpap) [inline examples/integration/widgets/data_table_customization.py]: There is already `bokeh.colors.color.RGB.brightness()`, which uses a variation of this method. bokehjs has one slightly different as well. We need a well-defined and consistent API for this.

File: bokehjs/src/lib/models/widgets/tables/cell_formatters.ts (hunk 6 above)
Comment (thomktz) [reply bokehjs/src/lib/models/widgets/tables/cell_formatters.ts:86]: I will probably need guidance on this - not sure where/when to precompute. Are we doing something similar somewhere else in the code I can use for inspiration ?

File: examples/integration/widgets/data_table_customization.py (hunk 9 above)
Comment (thomktz) [reply examples/integration/widgets/data_table_customization.py]: I think that luminance is a better contrast metric than brightness, so I added it as `bokeh.colors.color.RGB.luminance`, and added the bokehJS equivalent too.

Now looks cleaner:

```python
'#000000' if RGB.from_hex_string(hex_color).luminance > 0.4 else '#ffffff'
```

File: bokehjs/src/lib/models/widgets/tables/cell_formatters.ts (hunk 6 above)
Comment (bryevdv) [reply bokehjs/src/lib/models/widgets/tables/cell_formatters.ts:86]: @mattpap is this something we could merge as-is for now with a follow-up issue to improve later?

File: src/bokeh/colors/color.py [hunk 10]
```diff
@@ -365,6 +365,13 @@ def brightness(self) -> float:
         r, g, b = self.r, self.g, self.b
         return sqrt(0.299*r**2 + 0.587*g**2 + 0.114*b**2)/255
 
+    @property
+    def luminance(self) -> float:
+        """ Perceived luminance of a color in [0, 1] range. """
+        # https://en.wikipedia.org/wiki/Relative_luminance
+        r, g, b = self.r, self.g, self.b
+        return (0.2126*r**2.2 + 0.7152*g**2.2 + 0.0722*b**2.2) / 255**2.2
```
Comment (mosc9575) [inline src/bokeh/colors/color.py:373]:
```suggestion
        return sqrt(0.299*self.r**2 + 0.587*self.g**2 + 0.114*self.b**2)/255

    @property
    def luminance(self) -> float:
        """ Perceived luminance of a color in [0, 1] range. """
        # https://en.wikipedia.org/wiki/Relative_luminance
        return (0.2126*self.r**2.2 + 0.7152*self.g**2.2 + 0.0722*self.b**2.2) / 255**2.2
```
I think we don't need extra variables.

Comment (thomktz) [reply src/bokeh/colors/color.py:373]: Agree, did the change for brightness too

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
@thomktz this LGTM but let's wait for @mattpap to chime in about the one remaining question (i.e. change here with guidance, or make an issue for follow-up later)

File: src/bokeh/colors/color.py [hunk 11]
```diff
@@ -362,8 +362,13 @@ def to_rgb(self) -> RGB:
     def brightness(self) -> float:
         """ Perceived brightness of a color in [0, 1] range. """
         # http://alienryderflex.com/hsp.html
-        r, g, b = self.r, self.g, self.b
-        return sqrt(0.299*r**2 + 0.587*g**2 + 0.114*b**2)/255
+        return sqrt(0.299*self.r**2 + 0.587*self.g**2 + 0.114*self.b**2) / 255
```
Comment (mattpap) [inline src/bokeh/colors/color.py]: This seems to be an unnecessary change and it doesn't improve readability.

File: bokehjs/src/lib/models/widgets/tables/cell_formatters.ts [hunk 12]
```diff
@@ -56,37 +59,96 @@ export class StringFormatter extends CellFormatter {
...
+      if (text_color.value != null) {
+        text.style.color = color2css(text_color.value)
+      }
+    } else if (isField(text_color)) {
+      if (text_color.transform != null && text_color.transform instanceof ColorMapper) {
+        const rgbaArray = text_color.transform.rgba_mapper.v_compute([_dataContext[text_color.field]])
+        const [r, g, b, a] = rgbaArray
+        text.style.color = rgba2css([r, g, b, a])
+      } else {
+        text.style.color = _dataContext[text_color.field]
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/tables/cell_formatters.ts]: A data point in a color array can be almost anything, so this should also go through `color2css()` as well.

Comment (mattpap) [reply bokehjs/src/lib/models/widgets/tables/cell_formatters.ts]: Same applies to `background_color`.

File: bokehjs/src/lib/models/widgets/tables/cell_formatters.ts [hunk 13]
```diff
@@ -56,37 +59,96 @@ export class StringFormatter extends CellFormatter {
...
+
+    // Text color
+    // Handle the most common case first : isValue and value == null
+    if (isValue(text_color)) {
+      if (text_color.value != null) {
+        text.style.color = color2css(text_color.value)
+      }
+    } else if (isField(text_color)) {
+      if (text_color.transform != null && text_color.transform instanceof ColorMapper) {
+        const rgbaArray = text_color.transform.rgba_mapper.v_compute([_dataContext[text_color.field]])
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/tables/cell_formatters.ts]: This also needs renaming to camel case.

File: bokehjs/src/lib/models/widgets/tables/cell_formatters.ts [hunk 14]
```diff
@@ -56,37 +59,96 @@ export class StringFormatter extends CellFormatter {
...
 
     const text = div(value == null ? "" : `${value}`)
-    switch (font_style) {
+
+    // Font style
+    let resolved_font_style
+    if (isValue(font_style)) {
+      resolved_font_style = font_style.value
+    } else if (isField(font_style)) {
+      resolved_font_style = _dataContext[font_style.field]
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/tables/cell_formatters.ts]: `_` prefix indicates unused variable/argument and affects linting. If used, then it should be removed.

File: bokehjs/src/lib/models/widgets/tables/cell_formatters.ts (hunk 6 above)
Comment (mattpap) [reply bokehjs/src/lib/models/widgets/tables/cell_formatters.ts:86]: This is definitely not ideal and compiles only because it's mostly dynamically typed code, but given the proper solution is non-trivial and that this isn't critical code path in bokehjs, then yes we can proceed with this and improve later.

File: bokehjs/src/lib/models/widgets/tables/cell_formatters.ts [hunk 15]
```diff
@@ -56,37 +59,96 @@ export class StringFormatter extends CellFormatter {
...
-      text.style.color = color2css(text_color)
+    // Background color
+    if (isValue(background_color)) {
+      if (background_color.value != null) {
+        text.style.backgroundColor = color2css(background_color.value)
+      }
+    } else if (isField(background_color)) {
+      if (background_color.transform != null && background_color.transform instanceof ColorMapper) {
+        const rgbaArray = background_color.transform.rgba_mapper.v_compute([_dataContext[background_color.field]])
+        const [r, g, b, a] = rgbaArray
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/tables/cell_formatters.ts]: This is a bit confusing at the first glance, but given that `v_compute` returns a flat array (technically an ndarray), then this is indeed correct. Sadly there is no scalar variant in `ColorMapper`, so we will have to keep this as-is.

**Review by mattpap — APPROVED** [review]
Not ideal, but it works and we can improve this later. Please address the remaining comments, especially about missing `color2css()`.

**Comment by bryevdv** [comment]
Thanks @thomktz ! 🎉
